frameworks/TunedRandomForest/exec.py

```diff
@@ -64,60 +80,112 @@ def run(dataset, config):
-        tuning_durations = []
-        for i, max_features_value in enumerate(max_features_values):
-            log.info("[{:2d}/{:2d}] Evaluating max_features={}"
-                     .format(i + 1, len(max_features_values), max_features_value))
-            imputation = SimpleImputer()
-            random_forest = estimator(n_jobs=n_jobs,
-                                      random_state=config.seed,
-                                      max_features=max_features_value,
-                                      **tuning_params)
-            pipeline = Pipeline(steps=[
-                ('preprocessing', imputation),
-                ('learning', random_forest)
-            ])
-            with Timer() as cv_scoring:
+
+    log.info("Evaluating multiple values for `max_features`: %s.", max_features_values)
+    max_feature_scores = defaultdict(list)
+    tuning_durations = defaultdict(list)
+    memory_usage_by = defaultdict(list)
+    this_process = psutil.Process(os.getpid())
+    last_initial_fit_time = 0
+
+    with Timer() as training:
+        while max_features_values:
+            time_left = (config.max_runtime_seconds - training.duration)
+            time_per_value = time_left / (len(max_features_values) + 1)
+            if time_per_value < last_initial_fit_time:
+                log.info("Expect to exceed time constraints on next first fit, "
+                         f"budget is {time_per_value}s and last first fit took "
+                         f"{last_initial_fit_time}s.")
+                log.info(f"Did not try max_features={max_features_values}.")
+                break
+
+            value = max_features_values.pop(0)
+            log.info(f"Evaluating max_features={value} in {time_per_value} seconds.")
+
+            random_forest = estimator(
+                n_jobs=n_jobs,
+                random_state=config.seed,
+                max_features=value,
+                n_estimators=step_size,
+                warm_start=True,
+                **training_params
+            )
+
+            training_times = [training.duration]
+            memory_usage = [this_process.memory_info()[0] / (2 ** 20)]
+
+            while True:
                 try:
-                    scores = cross_val_score(estimator=pipeline,
-                                             X=X_train,
-                                             y=y_train,
-                                             scoring=metric,
-                                             error_score='raise',
-                                             cv=5)
-                    max_feature_scores.append((statistics.mean(scores), max_features_value))
-                except stopit.utils.TimeoutException as toe:
-                    log.error("Failed CV scoring for max_features=%s : Timeout", max_features_value)
-                    tuning_durations.append((max_features_value, cv_scoring.duration))
-                    raise toe
+                    scores = cross_val_score(
```

Comment:
What's the behavior here? Will cross_val_score simply call random_forest.fit(fold_X, fold_y)? Because in that case, because of warm-starting, will it train fold 2 on the ensemble pre-trained on fold 1? Or is it smart enough to build 5 fresh models? If so, and you run it in a loop (as done here), will it correctly warm-start the 5 individual models? Or start fresh again?

Reply:
> Or is it smart enough to build 5 fresh models? 

[Yes](https://github.com/scikit-learn/scikit-learn/blob/0d378913be6d7e485b792ea36e9268be31ed52d0/sklearn/model_selection/_validation.py#L269), but (as you point out) it should not smart enough to use warm-start in a loop (for the same reason). I will write something to verify this. I'm not actually sure that there's a way to achieve this within scikit-learn `cross_validate`-like functions.

Reply:
It seems that `warm_start`  and `cross_validate` are indeed [incompatible](https://github.com/scikit-learn/scikit-learn/discussions/22042#discussioncomment-1852489) and forests are trained from scratch each time. Will have to change the code to instead perform our own cross validation. I opened [an issue](https://github.com/scikit-learn/scikit-learn/issues/22044) hoping that functionality will be added in the future.

Reply:
I added an additional module which contains modified scikit-learn code which now enables using `warm_start` together with `cross_validate` and updated the `tunedrandomforest` code accordingly.